openlr/openlr_decoder.cpp

```diff
@@ -50,6 +69,142 @@ struct alignas(kCacheLineSize) Stats
   uint32_t m_tightOffsets = 0;
   uint32_t m_total = 0;
 };
+
+void ExpandFakes(Index const & index, Graph & g, Graph::EdgeVector & path)
+{
+  auto const isRealVertex = [&index](FeatureID const & fid, m2::PointD const & p) {
+    Index::FeaturesLoaderGuard g(index, fid.m_mwmId);
+    auto const ft = g.GetOriginalFeatureByIndex(fid.m_index);
+    bool matched = false;
+    ft->ForEachPoint(
+        [&p, &matched](m2::PointD const & fp) {
+          if (p == fp)
+            matched = true;
+        },
+        FeatureType::BEST_GEOMETRY);
+    return matched;
+  };
+
+  auto const expandFake = [&g, &path, isRealVertex](Graph::EdgeVector::iterator edgeIt) {
+    if (!edgeIt->IsFake())
+      return;
+
+    Graph::EdgeVector edges;
+    if (isRealVertex(edgeIt->GetFeatureId(), edgeIt->GetStartPoint()))
+      g.GetRegularOutgoingEdges(edgeIt->GetStartJunction(), edges);
+    else
+      g.GetRegularIngoingEdges(edgeIt->GetEndJunction(), edges);
+
+    CHECK(!edges.empty(), ());
+
+    auto const it = find_if(begin(edges), end(edges), [&edgeIt](Graph::Edge const & real) {
+      if (real.GetFeatureId() == edgeIt->GetFeatureId() && real.GetSegId() == edgeIt->GetSegId())
+        return true;
+      return false;
+    });
+
+    CHECK(it != end(edges), ());
+
+    if (2 * EdgeLength(*edgeIt) >= EdgeLength(*it))
+      *edgeIt = *it;
+    else
+      path.erase(edgeIt);
+
+    LOG(LDEBUG, ("len(e):", EdgeLength(*edgeIt), "len(*it):", EdgeLength(*it)));
+  };
+
+  expandFake(begin(path));
```

Comment:
ASSERT that first and last are fakes?
ASSERT that path is non-empty?

Reply:
> ASSERT that first and last are fakes? 

No need to ExpandFake will handle it and leave an edge as is if it's not fake.

Reply:
> ASSERT that path is non-empty> 

Done.